### PR TITLE
Fixed the translated3D Z value for each row to support dropdown in a …

### LINKED ISSUE #1180: Transform translate3D blocks dropdown in a cell
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, post on Stackoverflow or Gitter
```

**Current behavior**
If you try using a drop-down in any of the cells while **scrollbarV** is true and each row has a height the drop-down will be hidden under the next row.
An example is shown here - https://jsfiddle.net/sanborn/ejxregnc/

**Expected behavior**

The dropdown should be visible in that case. 
A solution example is shown here - https://jsfiddle.net/sanborn/ejxregnc/5/

**Reproduction of the problem**

Just try adding a drop-down in any of the cells.

**What is the motivation / use case for changing the behavior?**

We may need a drop-down menu for each row.

**Please tell us about your environment:**

* **Table version:** 0.8.x

* **Angular version:** 2.0.x

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]

### PR DESCRIPTION
**Pull request by sanbornsen**
…cell (fixes #1180)

**What kind of change does this PR introduce?** (check one with "x")
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Other... Please describe: Support for dropdowns in cell

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/swimlane/ngx-datatable/issues/1180

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: src/components/body/body.component.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ import { MouseEvent } from '../../events';
         [scrollbarH]="scrollbarH"
         [scrollHeight]="scrollHeight"
         [scrollWidth]="columnGroupWidths?.total"
+        [ngStyle]="{'transform-style': 'preserve-3d'}"
```
Comment (amcdnl) [inline src/components/body/body.component.ts]: No reason to make this `ngStyle`, just do a normal inline style.

File: src/components/body/body.component.ts [hunk 2]
```diff
@@ -485,19 +485,20 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
...
-      }        
+      }
 
       // const pos = idx * rowHeight;
       // The position of this row would be the sum of all row heights
       // until the previous row position.
       const pos = this.rowHeightsCache.query(idx - 1);
 
-      translateXY(styles, 0, pos);
+      // Z position of trnaslate3d
```
Comment (wizarrc) [inline src/components/body/body.component.ts]: spelling error

**Comment by wizarrc** [comment]
@sanbornsen This looks like in interesting fix. Can you provide a demo of how this feature works? Does this work with dropdown menus at the bottom of the view? This looks like it has a direction bias for an overlay menu.

In my scenario, say you want to right click which triggers a dropdown menu, but the menu would clip past the view. In many desktop programs, the dropdown menu would be responsive to the constraint in instead display starting at the top of the row instead of the bottom of the row when there is no space available.

**Comment by amcdnl** [comment]
I tend to agree w/ @wizarrc - typically you want to move your dropdown to the body and then position relatively for this very reason. Checkout - https://medium.com/@amcdnl/building-tooltips-for-angular2-396320fa938f

**Comment by wizarrc** [comment]
@amcdnl without a demo, I am skeptical this would even work if a drop down menu is implemented inside a row or cell template. My understanding of how clipping works is because it's inside the row element, it can't escape past the cell, despite the  `z-index`. I didn't know this restriction can be bypassed with `translate3d` css. A demo would help test my assumptions.

**Comment by sanbornsen** [comment]
I have added a basic demo. 

True that this solution will not work for dropup but only for dropdown.

File: src/utils/translate.ts [hunk 3]
```diff
@@ -9,13 +9,13 @@ const hasCSS3DTransforms = typeof window !== 'undefined' ? !!getVendorPrefixedNa
...
     if (!isSafari && hasCSS3DTransforms) {
-      styles[transform] = `translate3d(${x}px, ${y}px, 0)`;
+      styles[transform] = `translate3d(${x}px, ${y}px, ${z}px)`;
       styles[backfaceVisibility] = 'hidden';
     } else {
-      styles[camelCase(transform)] = `translate(${x}px, ${y}px)`;
+      styles[camelCase(transform)] = `translate(${x}px, ${y}px, ${z}px)`;
     }
   } else {
     styles.top = `${y}px`;
```
Comment (amcdnl) [inline src/utils/translate.ts]: What about support for top/left scenarios?

Comment (sanbornsen) [reply src/utils/translate.ts]: So the drop-down components usually have its own high z-index value. Therefore, it should work for top/left scenarios.

**Comment by sanbornsen** [comment]
We are also adding the same feature for the header in a more elegant way. So hold on to this PR.

**Pull request closed without merging**